### DISCUSSION TITLE
all keybindings now occur on key up instead of key down

### DISCUSSION THREAD

**Discussion by cognitiond** [davatorium/rofi#1742] [(Possible) Issues]
I'm not sure if this is intentional, so I'm asking here, but commit 6d02648d3a2ab9ee9751db8783f00b14071f57b4 causes keybindings (like move up an entry, down an entry, select an entry, everything) to run on keyup rather than keydown as it was before.

**Comment by DaveDavenport** [comment]
That is not intentional. 
@duarm  can we fix this?

**Reply by DaveDavenport** [reply]
@cognitiond  I cannot reproduce this. for me I can set it to respond on both press as release and that seems to work.  Can you give more info about setup? (imdkit version, distro)

**Reply by cognitiond** [reply]
I'm on arch linux, rofi built on the 6d02648 commit, xcb-imdkit package version 1.0.3-1, which hasn't changed for at least a year.
So, on further testing - setting the repeat delay to thousands of ms, it doesn't seem to be on keyup, but instead after a variable delay after keydown, anywhere from ~100ms to ~2000ms, unless I let the key go before that time, in which case it fires on that keyup.
What do you mean by "set it to respond on both press as release"? I can try reproducing what you did

**Reply by cognitiond** [reply]
This screengrab might make it clearer; the key showing up at the bottom corresponds to me pressing the key down, and I only release it after rofi has changed entries (key repeat delay is set to 5000ms). As you can see, there is a pretty significant delay after keydown after which rofi does anything. However, if I release the key quickly, it changes on keyup (or just at the same time?).
https://user-images.githubusercontent.com/118043135/201908410-b59cbd3d-5cbe-4e84-88f9-b0ed313b9824.mp4

**Reply by DaveDavenport** [reply]
for me it behave correctly, on press it is instantly and inverse on release it is too..

But yeah that video aint great :(, first thing I will do is see if I can make a flag to disable it compile time for testing. 

I did see before on some machine that for the first key press there was a delay, I wondered if that was an 'initialization' setup time.  multiple keypresses afterwards where fast.

Do you have fcitx or similar running? 

@duarm  any feedback on this?

**Reply by DaveDavenport** [reply]
added compile time flag. Can you test by disabling imdkit and see if that fixes it?

**Reply by cognitiond** [reply]
> Do you have fcitx or similar running?

I did in fact have fcitx running, and running rofi without fcitx running seems to fix it! Running rofi compiled with imdkit disabled also fixes the problem (with fcitx running). So, the problem only appears with imdkit enabled and fcitx running. Not sure if it matters, but I'm running fcitx5 version 5.0.19.
Also, would you mind if create a pull request expanding on the documentation for building (specifically, providing options to meson)? I had a bit of trouble figuring out how to actually pass the option to disable imdkit to meson.

**Reply by cognitiond** [reply]
Oh and another thing, do you have matrix, and if so, is there any matrix room for rofi?

**Reply by DaveDavenport** [reply]
PR for documentation improvements are always welcome.

For matrix I took the blue pill, but we have an irc channel: https://github.com/davatorium/rofi#discussion-places

**Reply by cognitiond** [reply]
Typing in the prompt is even worse: 
[input-lag.webm](https://user-images.githubusercontent.com/118043135/202825107-749237ba-2958-4311-841c-4f0c4d4d2f2d.webm)

**Reply by DaveDavenport** [reply]
I get no way to view that video, but I've already updated git to disable it by default.
I hope the writer of the patch can fix this.

On my 22.04 ubuntu machine I notice (and measure) no slowdown, it adds less then a ms to the key handling.
But apparently it is differs alot?  what distro/imdkit version do you use?

**Reply by DaveDavenport** [reply]
https://github.com/davatorium/rofi/pull/1735

**Reply by cognitiond** [reply]
I updated the video, hopefully this one works.
If it helps, this is what seems to be the relevant part of the timing trace:
```
4.090123 (0.000006): ../source/view.c:rofi_view_update:1236
4.090133 (0.000010): ../source/view.c:rofi_view_update:1254 Background
4.113370 (0.023237): ../source/view.c:rofi_view_update:1268 widgets
4.113396 (0.000026): ../source/view.c:rofi_view_repaint:300 Update start
4.113404 (0.000008): ../source/view.c:rofi_view_repaint:303 Expose
4.113414 (0.000010): ../source/view.c:rofi_view_repaint:308 flush
5.513041 (1.399627): ../source/view.c:rofi_view_refilter_real:1292 Filter start
5.513069 (0.000028): ../source/view.c:rofi_view_refilter_real:1297 Filter reload rows
5.513082 (0.000013): ../source/view.c:rofi_view_refilter_real:1302 Filter tokenize
5.513148 (0.000066): ../source/view.c:rofi_view_refilter_real:1381 Filter matching done
5.513157 (0.000009): ../source/widgets/listview.c:listview_set_num_elements:566 listview_set_num_elements
5.513164 (0.000007): ../source/widgets/listview.c:listview_set_num_elements:572 Set selected
5.513275 (0.000111): ../source/widgets/listview.c:listview_set_num_elements:574 recompute elements
5.513283 (0.000008): ../source/widgets/listview.c:listview_set_num_elements:576 queue redraw
5.513569 (0.000286): ../source/view.c:rofi_view_refilter_real:1394 Update filter lines
5.513626 (0.000057): ../source/view.c:rofi_view_refilter_real:1412 Filter resize window based on window
5.513635 (0.000009): ../source/view.c:rofi_view_refilter_real:1414 Filter done
```

Is the number in the parenthesis a delta from the last TICK_N() call?
